sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ExtRefService.java

```diff
@@ -5,41 +5,95 @@
 package org.lfenergy.compas.sct.commons.scl;
 
 import lombok.experimental.UtilityClass;
+import org.apache.commons.lang3.StringUtils;
 import org.lfenergy.compas.scl2007b4.model.SCL;
+import org.lfenergy.compas.scl2007b4.model.TCompasICDHeader;
+import org.lfenergy.compas.scl2007b4.model.TIED;
 import org.lfenergy.compas.sct.commons.dto.SclReport;
+import org.lfenergy.compas.sct.commons.dto.SclReportItem;
 import org.lfenergy.compas.sct.commons.scl.ied.IEDAdapter;
-import org.lfenergy.compas.sct.commons.scl.ied.InputsAdapter;
 import org.lfenergy.compas.sct.commons.scl.ied.LDeviceAdapter;
 import org.lfenergy.compas.sct.commons.scl.ied.LN0Adapter;
+import org.lfenergy.compas.sct.commons.util.PrivateEnum;
 
-import java.util.List;
+import java.util.*;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 
 @UtilityClass
 public class ExtRefService {
 
     /**
      * Updates iedName attribute of all ExtRefs in the Scd.
+     *
      * @return list of encountered errors
      */
     public static SclReport updateAllExtRefIedNames(SCL scd) {
         SclRootAdapter sclRootAdapter = new SclRootAdapter(scd);
+        List<SclReportItem> iedErrors = validateIed(sclRootAdapter);
+        if (!iedErrors.isEmpty()) {
+            return new SclReport(sclRootAdapter, iedErrors);
+        }
+        Map<String, IEDAdapter> icdSystemVersionToIed = sclRootAdapter.streamIEDAdapters()
+            .collect(Collectors.toMap(
+                iedAdapter -> PrivateService.getCompasPrivate(iedAdapter.getCurrentElem(), TCompasICDHeader.class)
+                    .map(TCompasICDHeader::getICDSystemVersionUUID)
+                    .orElseThrow(), // Value presence is checked by method validateIed called above
+                Function.identity()
+            ));
 
-        List<SclReport.ErrorDescription> errorDescriptions = sclRootAdapter.streamIEDAdapters()
+        List<SclReportItem> extRefErrors = sclRootAdapter.streamIEDAdapters()
             .flatMap(IEDAdapter::streamLDeviceAdapters)
             .filter(LDeviceAdapter::hasLN0)
             .map(LDeviceAdapter::getLN0Adapter)
             .filter(LN0Adapter::hasInputs)
             .map(LN0Adapter::getInputsAdapter)
-            .map(InputsAdapter::updateAllExtRefIedNames)
-            .flatMap(List::stream)
-            .collect(Collectors.toList());
-        return SclReport.builder()
-            .sclRootAdapter(sclRootAdapter)
-            .errorDescriptionList(errorDescriptions).build();
+            .map(inputsAdapter -> inputsAdapter.updateAllExtRefIedNames(icdSystemVersionToIed))
+            .flatMap(List::stream).collect(Collectors.toList());
+
+        return new SclReport(sclRootAdapter, extRefErrors);
+    }
+
+    private static List<SclReportItem> validateIed(SclRootAdapter sclRootAdapter) {
+        List<SclReportItem> iedErrors = new ArrayList<>(checkIedUnityOfIcdSystemVersionUuid(sclRootAdapter));
```

Comment:
calling checkIedCompasIcdHeaderAttributs before checkIedUnityOfIcdSystemVersionUuid is better i think

Reply:
You are right.